### PR TITLE
global: delay the url_map building

### PR DESCRIPTION
**Pull request by krzysztof**
- Delays the url_map building until first resolve request.

### CODE REVIEW

**Comment by krzysztof** [comment]
@jirikuncar Will this introduce any problems ?

I need this to build url_map where the path or host may contain config variables, e.g.: https://github.com/lnielsen/invenio-jsonschemas/blob/jsonresolver/invenio_jsonschemas/jsonresolver.py#L34-L43

**Comment by jirikuncar** [comment]
@krzysztof it looks good. Please add label `BETTER`, sign your commit, add yourself to AUTHORS, and add a test-case that check this behavior (instantiate JSONResolver, check url_map to be None, resolve(), and url_map is not None). Thanks

## 

update: even better check that your plugin was not called during JSONResolver.**init**

**Comment by krzysztof** [comment]
~~Ready~~
~~Python 3 problems~~
Ready

File: jsonresolver/hookspec.py [hunk 1]
```diff
@@ -3,9 +3,9 @@
 # This file is part of jsonresolver
 # Copyright (C) 2015 CERN.
 #
-# jsonresolver is free software; you can redistribute it and/or modify
-# it under the terms of the Revised BSD License; see LICENSE file for
-# more details.
+# jsonresolver is free software; you can redistribute it and/or
```
Comment (jirikuncar) [inline jsonresolver/hookspec.py]: why did you modify the license?

File: tests/demo/raising_hook.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+# Copyright (C) 2015 CERN.
+#
+# jsonresolver is free software; you can redistribute it and/or
+# modify it under the terms of the Revised BSD License; see LICENSE
+# file for more details.
+
+"""Test for detecting the lazy evaluation of hook-registered JSON resolver."""
+
+import jsonresolver
+from .raising import ExecutionDetect
```
Comment (jirikuncar) [inline tests/demo/raising_hook.py]: it probably needs isort

File: tests/demo/raising_hook.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+from werkzeug.routing import Rule
+
+
+@jsonresolver.hookimpl
+def jsonresolver_loader(url_map):
+    """Load the raising plugin as a Rule to URL map."""
+    def endpoint():
+        """Raise an exception."""
+        raise ExecutionDetect
```
Comment (jirikuncar) [inline tests/demo/raising_hook.py]: you probably want this on top level

Comment (jirikuncar) [inline tests/demo/raising_hook.py]: and maybe you can rename it to `HookRegistrationDetected`

**Comment by krzysztof** [comment]
- Fixed isort
- Renamed exception
- added top-level endpoint

**Comment by krzysztof** [comment]
Wait, this test is still wrong...

**Comment by krzysztof** [comment]
Ok, this one should test the lazy hook registration properly.

File: tests/demo/raising_hook.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+def endpoint():
+    """Test endpoint."""
+    return {'test': 'test'}
+
+
+@jsonresolver.hookimpl
+def jsonresolver_loader(url_map):
+    """Load the raising plugin as a Rule to URL map."""
+    raise HookRegistrationDetected
+    url_map.add(
```
Comment (jirikuncar) [inline tests/demo/raising_hook.py]: you can remove the dead code

File: tests/demo/raising_hook.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+from werkzeug.routing import Rule
+
+import jsonresolver
+
+
+class HookRegistrationDetected(Exception):
+    """Raise this ``exception`` to detect when a hook is registered."""
+
+
+def endpoint():
```
Comment (jirikuncar) [inline tests/demo/raising_hook.py]: not needed

**Comment by jirikuncar** [comment]
@krzysztof it looks good now. Please remove dead code so we can :shipit: Thanks!

**Comment by krzysztof** [comment]
Removed dead code
